packages/dds/tree/src/feature-libraries/default-schema/schemaChecker.ts

```diff
@@ -0,0 +1,153 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { assert, unreachableCase } from "@fluidframework/core-utils/internal";
+import {
+	MapTree,
+	StoredSchemaCollection,
+	TreeFieldStoredSchema,
+	LeafNodeStoredSchema,
+	ObjectNodeStoredSchema,
+	MapNodeStoredSchema,
+	Multiplicity,
+} from "../../core/index.js";
+import { FullSchemaPolicy } from "../modular-schema/index.js";
+import { allowsValue } from "../valueUtilities.js";
+
+export const enum SchemaValidationErrors {
+	NoError,
+	Field_KindNotInSchemaPolicy,
+	Field_IncorrectMultiplicity,
+	Field_NodeTypeNotAllowed,
+	LeafNode_MissingValue,
+	LeafNode_InvalidValue,
+	LeafNode_FieldsNotAllowed,
+	ObjectNode_FieldCountMismatch,
+	ObjectNode_FieldNotInSchema,
+	Node_MissingSchema,
+	UnknownError,
+}
+
+export function isNodeInSchema(
+	node: MapTree,
+	nodeSchemaCollection: StoredSchemaCollection,
+	schemaPolicy: FullSchemaPolicy,
+): SchemaValidationErrors {
+	// Validate the schema declared by the node exists
+	const schema = nodeSchemaCollection.nodeSchema.get(node.type);
+	if (schema === undefined) {
+		return SchemaValidationErrors.Node_MissingSchema;
+	}
+
+	// Validate the node is well formed according to its schema
+
+	if (schema instanceof LeafNodeStoredSchema) {
+		if (node.value === undefined) {
+			return SchemaValidationErrors.LeafNode_MissingValue;
+		}
+		if (node.fields.size !== 0) {
+			return SchemaValidationErrors.LeafNode_FieldsNotAllowed;
+		}
+		if (!allowsValue(schema.leafValue, node.value)) {
+			return SchemaValidationErrors.LeafNode_InvalidValue;
+		}
+	} else if (schema instanceof ObjectNodeStoredSchema) {
+		if (node.fields.size !== schema.objectNodeFields.size) {
+			return SchemaValidationErrors.ObjectNode_FieldCountMismatch;
```

Comment:
This check is invalid: map tree as a format requires empty fields be omitted, and object nodes to allow empty fields if the fiel kind's multiplicity allows it (like sequence and optional do)

Reply:
Hmm not entirely sure I got how this is invalid. Two questions:
- The Map case is handled below, and that one doesn't care how many fields actually exist in the node, which at least partially matches "requres empty fields to be omitted". If an empty field is actually present (field key with no child nodes) in the node, should that make that node fail schema validation? Or empty fields _may_ be ommitted but it's also fine if they're present?
- In the Object case, if an optional field doesn't "exist" (i.e. the field key is not present in the node's fields) in the node, the node would still be considered in schema? The code right now expects the field to be there and have no children, so the condition is checking "is the number of defined fields (independent of their contents, at this point) correct?".

Reply:
I think I'm leaning towards having empty fields in a map node be a normal (non-asserted) failure in the validator, precisely for cases where we can create a "lower-level-primitives" tree like we do here with MapTree. Are there other cases which might also not enjoy the benefits of the assertions that prevent this scenario? For those, I think a "not in schema" response here might be better than a failed assert triggering potentially long after they did the wrong thing (adding an empty field to a map, or whatever invalid thing we want to handle). Thoughts?